sbi/src/api/tee_host.rs

```diff
@@ -160,19 +146,28 @@ pub fn get_vcpu_register_set(vmid: u64, index: u64) -> Result<RegisterSetLocatio
     Ok(RegisterSetLocation::from(raw as u32))
 }
 
-/// Adds a vCPU with ID `vcpu_id` to the guest `vmid`, registering `shared_page_addr` as the
-/// location of the vCPU's shared-memory state area.
+/// Adds a vCPU with ID `vcpu_id` to the guest `vmid`, using 'state_page_addr' to hold the vCPU state and
+/// registering `shared_page_addr` as the location of the vCPU's shared-memory state area.
 ///
 /// # Safety
 ///
 /// The caller must own the pages referenced by `shared_page_addr`, for the number of pages
 /// sufficient to hold the structure layout returned from `get_vcpu_mem_layout()`. Since memory
 /// within the shared-memory state area may be read or written by the TSM at any time, the caller
 /// must treat the memory as volatile for the lifetime of the TVM.
-pub unsafe fn add_vcpu(vmid: u64, vcpu_id: u64, shared_page_addr: u64) -> Result<()> {
+///
+/// The address `state_page_addr` must reference confidential memory pages. The caller must provide
+/// `TsmInfo::tvm_vcpu_state_pages` pages.
```

Comment:
Yes.

If you follow that code on the other side of Salus you'll find yourself in [`guest_add_vcpu`](https://github.com/rivosinc/salus/blob/main/src/vm.rs#L1210). 
From there, we arrive to [`create_guest_vm`](https://github.com/glg-rv/salus/blob/topic/vcpu_allocation/src/vm_pages.rs#L1117). When we arrive here, the only check we've done in the address passed is essentially that it is page aligned, something that happened when we annotated the raw address into a full `GuestPageAddr`.

Here is where things actually happen -- we're in `vm_pages.rs` so that's expected. You should notice that the function is divided in two parts. The first is where we try to get ranges of contiguous converted pages and exit with error early in case we don't.
And the other where we take these pages, assign them, and make them into SequentialPages to create data structures. Anything that handles SequentialPages can't fail.
This is due to the fact that if we fail _before_ the assign+SequentialPages, we would still be able to release those pages on variable drop. Once we assign them, or they pages are saved in a SequentialPages, we won't.

This is why we check early the memory requirements, and are the memory requirement you're asking about.

The fundamental check consist in `get_converted_pages`, that gets a range of memory that have been previously been converted. This is the check that it is confidential, because pages would have followed the protocol as described in `TsmConvertPages` (and the fence procedure of TLB invalidation).
This will also enforce the precise number of pages we'll receive (or return an error).

So it does error out if it isn't confidential or it's not big enough, and does not write anything in host memory in those cases.

Reply:
Thanks for that detailed evaluation. In that case, I'd move the 'state_page_addr' restriction in the comment to above the safety section as it isn't unsafe, from rust's perspective, to pass an arbitrary pointer to this function.

Reply:
Ah yes I will. I think in some past iteration was told to add the comment there. I didn't know what it was meant for the safety section there. Just doing what I was told. :-)